### PR TITLE
Cucumber expressions regression 323

### PR DESCRIPTION
**Pull request by aslakhellesoy**
WIP fix for #326 

- [x] Java
- [x] Ruby
- [x] JavaScript

### CODE REVIEW

**Comment by aslakhellesoy** [comment]
I added the same tests to Ruby and JavaScript (locally, without committing), and they both passed without any changes!

I then changed the implementation in both Ruby and JavaScript to be aligned with the changes to the Java implementation. The JavaScript implementation still passed, but the Ruby implementation now failed!

That worries me a little. I don't want the implementations to diverge, since that would make maintenance harder.

I haven't tried to understand why the implementations would behave differently. Any ideas?

**Comment by aslakhellesoy** [comment]
I'm investigating this a little further. Let's look at the source of the regular expressions generated by the Cucumber Expression `three {string} and {string} mice`:

```
Java:
^three ("([^"\\]*(\\.[^"\\]*)*)"|\'([^\'\\]*(\\.[^\'\\]*)*)\') and ("([^"\\]*(\\.[^"\\]*)*)"|\'([^\'\\]*(\\.[^\'\\]*)*)\') mice$

JavaScript:
^three ("([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)') and ("([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)') mice$

Ruby:
^three ("([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)') and ("([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)') mice$
```

The JavaScript and Ruby regexps are identical, but the Java one differs in that `'` characters are escaped.

Maybe related - I [discovered today](https://github.com/cucumber/cucumber/issues/325#issuecomment-358985126) that Ruby and JavaScript escapes `/` differently in regexp source.

Tricky stuff!

**Comment by mpkorstanje** [comment]
The `\'` is in the java defintion of `{string}`.

**Comment by mpkorstanje** [comment]
My fix is definitely wrong.

**Comment by aslakhellesoy** [comment]
Looks like changing `\\'` to `'` in the `{string}` regexp fixes it! And no changes needed to Ruby/JavaScript because they don't have it!

**Comment by mpkorstanje** [comment]
That should do it. `'` is not a character that would have to be escaped.

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeRegistry.java [hunk 1]
```diff
@@ -14,8 +14,7 @@
     private static final List<String> FLOAT_REGEXPS = singletonList("-?\\d*[\\.,]\\d+");
     private static final List<String> HEX_REGEXPS = singletonList("0[xX][0-9a-fA-F]{2}");
     private static final List<String> WORD_REGEXPS = singletonList("\\w+");
-    private static final List<String> STRING_REGEXPS = singletonList("\"([^\"\\\\]*(\\\\.[^\"\\\\]*)*)\"|\\'([^\\'\\\\]*(\\\\.[^\\'\\\\]*)*)\\'");
-
+    private static final List<String> STRING_REGEXPS = singletonList("\"([^\"\\\\]*(\\\\.[^\"\\\\]*)*)\"|'([^'\\\\]*(\\\\.[^'\\\\]*)*)'");
```
Comment (mpkorstanje) [inline cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeRegistry.java]: I think I did this right.

**Comment by mpkorstanje** [comment]
Eh there is a bit in the ParameterTypeRegistry too that needs to change. Have to look at it another time.

**Comment by aslakhellesoy** [comment]
I fixed the missing bits (I think!)

**Pull request merged**